paper/paper.md

```diff
@@ -0,0 +1,103 @@
+The Causal Testing Framework is a software testing framework that uses Causal Inference techniques to establish causal effects between software variables from pre-existing runtime data rather than having to collect bespoke, highly curated datasets especially for testing.
+
+# Statement of need
+Metamorphic Testing [@chen1998metamorphic] is a popular technique for testing computational models (and other traditionally "hard to test" software).
+Test goals are expressed as _metamorphic relations_ that specify how changing an input in a particular way should affect the software output.
+Nondeterministic software can be tested using Statistical Metamorphic Testing [@guderlei2007smt], which uses statistical tests over multiple executions of the software to determine whether the specified metamorphic relations hold.
+However, this requires the software to be executed repeatedly for each set of parameters of interest, so is computationally expensive, and is constrained to testing properties over software inputs that can be directly and precisely controlled.
+Statistical Metamorphic Testing cannot be used to test properties that relate internal variables or outputs to each other, since these cannot be controlled a priori.
+
+By employing domain knowledge in the form of a causal graph --- a lightweight model specifying the expected relationships between key software variables --- the Causal Testing Framework circumvents both of these problems by enabling models to be tested using pre-existing runtime data.
+The Causal Testing Framework is written in python but is language agnostic in terms of the system under test.
+All that is required is a set of properties to be validated, a causal model, and a set of software runtime data.
+
+# Causal Testing
+Causal Testing [@clark2023testing] has four main steps, outlined in \ref{fig:schematic}.
+Firstly, the user supplies a causal model, which takes the form of a directed acyclic graph (DAG) in which an edge $X \to Y$ represents variable $X$ having a direct causal effect on variable $Y$.
+Secondly, the user supplies a set of causal properties to be tested.
+Such properties can be generated from the causal DAG [@clark2023metamorphic]: for each $X \to Y$ edge, a test to validate the presence of a causal effect is generated, and for each missing edge, a test to validate independence is generated.
+The user may also refine tests to validate the nature of a particular relationship.
+Next, the user supplies a set of runtime data in the form of a table with each column representing a variable and rows containing the value of each variable for a particular run of the software.
+Finally, the Causal Testing Framework automatically validates the supplied causal properties by using the supplied causal DAG and data to calculate a causal effect estimate, and validating this against the expected causal relationship.
```

Comment:
@jmafoster1 "...validates the supplied causal properties by using the supplied.." remove one or both instances of "supplied" here?